app/code/community/Sendsmaily/Sync/Model/Observer.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+/**
+ * Sendsmaily Sync
+ * Export Magento newsletter subscribers to Sendsmaily
+ * Copyright (C) 2010 Sendsmaily
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+class Sendsmaily_Sync_Model_Observer
+{
+    /**
+     * Observers newsletter subscribe form.
+     *
+     * @param Varien_Event_Observer $observer
+     * @return void
+     */
+    public function newsletterOptInSubscribe($observer)
+    {
+        // Run only if newsletter subscriber opt-in is enabled.
+        if (!Mage::helper('sync')->newsletterOptInEnabled()) {
+            return;
+        }
+
+        // TODO: Should run if email is allready in newsletter subscribers list?
```

Comment:
Magento checks if [email exists in customer model only](https://github.com/OpenMage/magento-mirror/blob/af2c70ff9f4fec9373595689d48180b3808893cc/app/code/core/Mage/Newsletter/controllers/SubscriberController.php#L56-L63)
```
$ownerId = Mage::getModel('customer/customer')
   ->setWebsiteId(Mage::app()->getStore()->getWebsiteId())
   ->loadByEmail($email)
   ->getId();
if ($ownerId !== null && $ownerId != $customerSession->getId()) {
  Mage::throwException($this->__('This email address is already assigned to another user.'));
}
```
It does not check the newsletter subscribers' collection for a match. It just subscribes the same email address again. So basically yes, you can trigger our opt-in with the same email address that is not assigned to a customer.

Reply:
Okay, so I found a logic mistake. Magento runs our observer even if it had some errors thrown in the Subscriber controller. The subscriber controller just catches it's own errors and writes them to the session basically. So nothing stops our observer running.

I think we should check if subscribed email already exists in both Subscribers and Customers collection and if it does we omit opt-in to Smaily and don't send the no dispatch flag. As the main purpose of this feature is to send welcome emails I think those customers don't need it anyway.

Reply:
Okey this check can be improved. The new email should be new to newsletter subscribers list and if it is a match for a customer we should check if the current customer is the one signing up for the newsletter.

Reply:
Implemented this logic in db5f1a6025fc752c7739437d38b3644d68299912

Reply:
It is getting really complicated.

What happens when newsletter subscriber unsubscribes? Would he/she still be in the newsletter subscribers database, but in a different state? If so, then we should be able to handle the "unsubscribed to subscribed" state change also as an opt-in.

Maybe we can rethink the logic and use a different hook for that?

Reply:
>What happens when newsletter subscriber unsubscribes?

Yes, it will stay in the database and have an unsubscribed status.

>Maybe we can rethink the logic and use a different hook for that?

When I started with opt-in integration I was also thinking that maybe hooking into Subscriber model saving is a better idea. Only then we need the same logic to determine when to run captcha validation and when not. The model saving action can take place from the customers' setting page also and trigger the observer. I haven't found a way to determine which part of Magneto triggered this action. (Newsletter form or customer changing its subscription status) So I can't disable CAPTCHA checking if save is triggered form the customer settings page.

![image](https://user-images.githubusercontent.com/22918359/69861169-7aa62e80-12a0-11ea-8bf2-21b98a640183.png)


> If so, then we should be able to handle the "unsubscribed to subscribed" state change also as an opt-in.

To implement both I think we need to keep the current observer for CAPTCHA check and separate observer `customer_save_after` for checking if the customer has saved with subscribed status.

Reply:
I separated the CAPTCHA check and model change to separate observers. CAPTCHA check only checks if captcha is valid and kills controller if not so the other observer that controls opt-in and opt-out won't be triggered.

Model saving triggers opt-in/opt-out workflow if subscription status is changed.
https://github.com/sendsmaily/Sendsmaily-Sync-for-Magento/pull/7/commits/1ba57bc8862776ad4e23216f1ceb77c6b3871337